src/i2c/src/vhu_i2c.rs

```diff
@@ -194,7 +194,7 @@ impl<D: I2cDevice> VhostUserI2cBackend<D> {
             if desc_buf.is_write_only() {
                 desc_chain
                     .memory()
-                    .write(&reqs.remove(0).buf, desc_buf.addr())
+                    .write(&reqs[i].buf, desc_buf.addr())
```

Comment:
https://doc.rust-lang.org/std/vec/struct.Vec.html#method.remove

The existing code is fine because it removes the items from the vector one by one at index 0, which results in shifting of the rest of the stuff to left by 1 place.

Reply:
Not really, it would be fine to do it the way suggested here, just need to fix the commit log to mention that it is an improvement and not a bugfix

Reply:
Actually, this commit does in fact fix incorrect data and guest memory corruptions.

The problem with the current code is that the remove() is called only on
chains with writeable buffers.  So I2C transactions where there is a write
followed by a read (a very common type of I2C transaction) will result in the
aforementioned problems.

Here are some logs.  Host system:

```
$ sudo modprobe i2c-stub chip_addr=0x20
$ i2cdump -y -r 0x00-0xf 7 0x20 b
     0  1  2  3  4  5  6  7  8  9  a  b  c  d  e  f    0123456789abcdef
00: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00    ................
$ i2cset -y 7 0x20 0x01 0xde
$ i2cset -y 7 0x20 0x02 0xad
$ i2cset -y 7 0x20 0x03 0xbe
$ i2cset -y 7 0x20 0x04 0xef
$ i2cdump -y -r 0x00-0xf 7 0x20 b
     0  1  2  3  4  5  6  7  8  9  a  b  c  d  e  f    0123456789abcdef
00: 00 de ad be ef 00 00 00 00 00 00 00 00 00 00 00    .????...........

$ target/debug/vhost-device-i2c -l 7:20 -s foo
...
```

Guest, v5.15-rc5 with virito-i2c, before this patch:

```
root@(none):~# i2cdump -y -r 0x00-0xf 0 0x20 b
     0  1  2  3  4  5  6  7  8  9  a  b  c  d  e  f    0123456789abcdef
00: 00 01 02 03 04 05 06 07 08 09 0a 0b 0c 0d 0e 0f    .???????????????
```

Guest, after this patch:

```
root@(none):~# i2cdump -y -r 0x00-0xf 0 0x20 b
     0  1  2  3  4  5  6  7  8  9  a  b  c  d  e  f    0123456789abcdef
00: 00 de ad be ef 00 00 00 00 00 00 00 00 00 00 00    .????...........
```

And the memory corruptions are also easily reproducible by using a transaction
which has a larger buffer followed by a shorter buffer:

```
root@(none):~# i2ctransfer -y 0 w1@0x20 42 r64 r32
[    6.210000][   T22] i2c_write: i2c-0 #0 a=020 f=0200 l=1 [2a]
[    6.210000][   T22] i2c_read: i2c-0 #1 a=020 f=0201 l=64
[    6.210000][   T22] i2c_read: i2c-0 #2 a=020 f=0201 l=32
[    6.210000][   T22] i2c_result: i2c-0 n=3 ret=0
[    6.210000][   T22] =============================================================================
[    6.210000][   T22] BUG kmalloc-32 (Not tainted): Right Redzone overwritten
[    6.210000][   T22] -----------------------------------------------------------------------------
[    6.210000][   T22] 
[    6.210000][   T22] 0x00000000627d18a0-0x00000000627d18a7 @offset=6304. First byte 0x0 instead of 0xcc
[    6.210000][   T22] Freed in load_elf_binary+0x18a/0xce1 age=553 cpu=0 pid=1
[    6.210000][   T22] 	save_stack_trace+0x33/0x35
[    6.210000][   T22] 	stack_trace_save+0x39/0x3e
[    6.210000][   T22] 	set_track.isra.0+0x4c/0xb8
[    6.210000][   T22] 	free_debug_processing+0xdd/0x30d
[    6.210000][   T22] 	__slab_free+0x6f/0x1ec
[    6.210000][   T22] 	kfree+0x1bd/0x1cc
[    6.210000][   T22] 	load_elf_binary+0x18a/0xce1
[    6.210000][   T22] 	exec_binprm+0x155/0x47a
[    6.210000][   T22] 	bprm_execve+0x1e8/0x2a9
[    6.210000][   T22] 	do_execveat_common+0x179/0x1b2
[    6.210000][   T22] 	sys_execve+0x3c/0x44
[    6.210000][   T22] 	handle_syscall+0xa9/0xd9
[    6.210000][   T22] 	userspace+0x4b4/0x542
[    6.210000][   T22] 	new_thread_handler+0xb5/0xb7
[    6.210000][   T22] Slab 0x00000000644d6580 objects=19 used=16 fp=0x00000000627d1a20 flags=0x10201(locked|slab|head|zone=0)
[    6.210000][   T22] Object 0x00000000627d1880 @offset=6272 fp=0x0000000000000000
[    6.210000][   T22] 
[    6.210000][   T22] Redzone  00000000627d1860: cc cc cc cc cc cc cc cc cc cc cc cc cc cc cc cc  ................
[    6.210000][   T22] Redzone  00000000627d1870: cc cc cc cc cc cc cc cc cc cc cc cc cc cc cc cc  ................
[    6.210000][   T22] Object   00000000627d1880: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00  ................
[    6.210000][   T22] Object   00000000627d1890: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00  ................
[    6.210000][   T22] Redzone  00000000627d18a0: 00 00 00 00 00 00 00 00                          ........
[    6.210000][   T22] Padding  00000000627d19e0: 5a 5a 5a 5a 5a 5a 5a 5a 5a 5a 5a 5a 5a 5a 5a 5a  ZZZZZZZZZZZZZZZZ
[    6.210000][   T22] Padding  00000000627d19f0: 5a 5a 5a 5a 5a 5a 5a 5a 5a 5a 5a 5a 5a 5a 5a 5a  ZZZZZZZZZZZZZZZZ
[    6.210000][   T22] CPU: 0 PID: 22 Comm: i2ctransfer Not tainted 5.15.0-rc5-00011-g398f504f9a8a-dirty #80
[    6.210000][   T22] Stack:
[    6.210000][   T22]  60611a1d 627d0000 6489fab0 61c02a40
[    6.210000][   T22]  627d1880 627d0000 6014556e 627d18a0
[    6.210000][   T22]  6489fac0 6049335a 6489faf0 60492a59
[    6.210000][   T22] Call Trace:
[    6.210000][   T22]  [<60491b4d>] ? _printk+0x0/0x99
[    6.210000][   T22]  [<60028983>] show_stack+0x143/0x152
[    6.210000][   T22]  [<6014556e>] ? print_section+0x0/0x31
[    6.210000][   T22]  [<6049335a>] dump_stack+0x39/0x3b
[    6.210000][   T22]  [<60492a59>] print_trailer+0x189/0x192
[    6.210000][   T22]  [<60147b8d>] check_bytes_and_report+0x112/0x17a
[    6.210000][   T22]  [<60147db2>] check_object+0xa0/0x23c
[    6.210000][   T22]  [<60149fd6>] free_debug_processing+0xa1/0x30d
[    6.210000][   T22]  [<6037f2f2>] ? i2cdev_ioctl_rdwr+0x1bb/0x1de
[    6.210000][   T22]  [<6014a242>] ? __slab_free+0x0/0x1ec
[    6.210000][   T22]  [<60145443>] ? arch_local_irq_save+0x0/0x29
[    6.210000][   T22]  [<6014a2b1>] __slab_free+0x6f/0x1ec
[    6.210000][   T22]  [<6037f2f2>] ? i2cdev_ioctl_rdwr+0x1bb/0x1de
[    6.210000][   T22]  [<60146230>] ? slab_free_freelist_hook+0x6c/0xf5
[    6.210000][   T22]  [<6009e49f>] ? rcu_read_lock_sched_held+0x0/0x39
[    6.210000][   T22]  [<60145443>] ? arch_local_irq_save+0x0/0x29
[    6.210000][   T22]  [<6014a5eb>] kfree+0x1bd/0x1cc
[    6.210000][   T22]  [<6037f2f2>] ? i2cdev_ioctl_rdwr+0x1bb/0x1de
[    6.210000][   T22]  [<6014a42e>] ? kfree+0x0/0x1cc
[    6.210000][   T22]  [<6037f2f2>] i2cdev_ioctl_rdwr+0x1bb/0x1de
[    6.210000][   T22]  [<6037f55c>] i2cdev_ioctl+0x247/0x2ed
[    6.210000][   T22]  [<6016396e>] vfs_ioctl+0x21/0x30
[    6.210000][   T22]  [<60164be7>] sys_ioctl+0xb18/0xb41
[    6.210000][   T22]  [<6002aed7>] ? segv+0x112/0x2d8
[    6.210000][   T22]  [<600281d3>] ? syscall_trace_enter+0x0/0x35
[    6.210000][   T22]  [<6002bc39>] handle_syscall+0xa9/0xd9
[    6.210000][   T22]  [<600426d5>] userspace+0x4b4/0x542
[    6.210000][   T22]  [<60027509>] fork_handler+0x99/0x9b
[    6.210000][   T22] Disabling lock debugging due to kernel taint
[    6.210000][   T22] FIX kmalloc-32: Restoring Right Redzone 0x00000000627d18a0-0x00000000627d18a7=0xcc
[    6.210000][   T22] FIX kmalloc-32: Object at 0x00000000627d1880 not freed
Warning: only 0/3 messages were sent
root@(none):~#
```

Reply:
@vitkyrka Yes, that's correct. Thanks for fixing it.